guides/api-guideline.md

```diff
@@ -0,0 +1,135 @@
+Next to that the guidelines defines a syntax for documenting the APIs in a language-agnostic way for proposals.
+
+## Language Agnostic API Documentation Syntax
+
+The [proposals](../proposals) folder contains design documents of new features or changes for our SDKs.
+Each proposal should include a section that documents the API in a language-agnostic way.
+
+### Basic data types
+
+The following basic data types should be used in the API documentation.
+
+| Data Type       | Description                                                   |
+|-----------------|---------------------------------------------------------------|
+| `string`        | A sequence of characters                                      |
+| `intX`          | A signed integer of X bits                                    |
+| `uintX`         | An unsigned integer of X bits                                 |
+| `int`           | A signed integer of 32 bits                                   |
+| `long`          | A signed integer of 64 bits                                   |
+| `double`        | A native floating-point number in 64-bit base-2 format        |
+| `decimal`       | A decimal number with arbitrary precision                     |
+| `bool`          | A boolean value                                               |
+| `bytes`         | A sequence of bytes                                           |
+| `list<TYPE>`    | A list of elements of type TYPE                               |
+| `set<TYPE>`     | A set of elements of type TYPE                                |
+| `map<TYPE>`     | An immutable map of elements of type TYPE                     |
+| `date`          | A date value (ISO 8601 calendar date)                         |
+| `time`          | A time value without date or timezone (nanosecond precision)  |
+| `dateTime`      | A date and time value without timezone (nanosecond precision) |
+| `zonedDateTime` | A date and time value with timezone (nanosecond precision)    |
+
+### Enumerations
+
+Enumerations can be defined using the following syntax:
+```
+enum EnumName {
+    VALUE1
+    VALUE2
+}
+```
+
+### Attribute annotations
+
+Attribute annotations can be used to provide additional information about attributes in complex data types.
+The following annotations should be used:
+- `@immutable`: Indicates that the field is immutable and cannot be changed after creation.
+- `@optional`: Indicates that the field is optional and can be null or undefined.
+- `@default(value)`: Indicates that the field has a default value.
+- `@deprecated`: Indicates that the field is deprecated and should not be used.
+- `@removed`: Indicates that the field must be removed.
+- `@min(value)`: Indicates the minimum value for numeric fields.
+- `@max(value)`: Indicates the maximum value for numeric fields.
+- `@minLength(value)`: Indicates the minimum length for string fields.
+- `@maxLength(value)`: Indicates the maximum length for string fields.
```

Comment:
Does the behavior need to specified if a value is provided lower than the min/above the max? Or should it not even be allowed?

Reply:
Should this be enforced at the SDK level or consensus node level? For example, memos are currently limited to 100 bytes. Enforcing at SDK level would prevent users from ever inputting >100 byte memos and make it fail faster. However, what if consensus node decides to up this to 150 bytes? Users would need to update SDKs to get this new enforcement. Whereas if we just allowed enforcement at the consensus node level always, SDKs wouldn't need to bother (and descriptive `MEMO_TOO_LONG` response codes would still highlight the exact issue).